tools/python/util/android/android.py

```diff
@@ -204,14 +214,126 @@ def start_emulator(
 
             _log.debug(f"sys.boot_completed='{getprop_value}'. Sleeping for {sleep_interval_seconds} before retrying.")
             time.sleep(sleep_interval_seconds)
-
+        # Verify if the emulator is now running
+        if not is_emulator_running_by_avd(avd_name=avd_name):
+            raise RuntimeError("Emulator failed to start.")
         return emulator_process
 
 
-def stop_emulator(emulator_proc_or_pid: typing.Union[subprocess.Popen, int]):
+def is_emulator_running_by_avd(avd_name: str) -> bool:
+    """
+    Check if an emulator is running based on the provided AVD name.
+    :param avd_name: Name of the Android Virtual Device (AVD) to check.
+    :return: True if an emulator with the given AVD name is running, False otherwise.
+    """
+    try:
+        # Step 1: List running devices
+        result = subprocess.check_output(["adb", "devices"], text=True).strip()
+        _log.info(f"adb devices output:\n{result}")
+        running_emulators = [line.split("\t")[0] for line in result.splitlines()[1:] if "emulator" in line]
+
+        if not running_emulators:
+            _log.debug("No emulators running.")
+            return False  # No emulators running
+
+        # Step 2: Check each running emulator's AVD name
```

Comment:
Should we have step 3 of checking `getprop sys.boot_completed` like we do in `start_emulator` to make sure they have successfully started?

Reply:
I don't think it is necessary, because It is used by `start_emulator` which calls this:
1.  at the beginning of the function to check if the emulator with the avd name is already running. 
2. at the end of the function to check if the emulator with the avd is running **AFTER** they have successfully started.